### PR TITLE
[GISEL] Add G_VSCALE instruction

### PR DESCRIPTION
**Pull request by michaelmaitland**

### CODE REVIEW

File: llvm/test/MachineVerifier/test_g_vscale.mir [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# RUN: not --crash llc -verify-machineinstrs -run-pass none -o /dev/null %s 2>&1 | FileCheck %s
+# REQUIRES: aarch64-registered-target
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_g_vscale.mir:2]: The target isn't specified here, and this doesn't need one as written

Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_vscale.mir]: Updated

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/Target/GenericOpcodes.td [hunk 2]
```diff
@@ -1289,6 +1289,15 @@ def G_MERGE_VALUES : GenericInstruction {
   let variadicOpsType = type1;
 }
 
+// Generic vscale.
+// Puts the value of the runtime vscale multiplied by the value in the source
+// operand into the destination register.
+def G_VSCALE : GenericInstruction {
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins untyped_imm_0:$src);
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/include/llvm/Target/GenericOpcodes.td:1297]: For G_CONSTANT, we retrieve the value as following:
https://github.com/llvm/llvm-project/blob/cf1319f9c6561afea381bbfc1a18f5c1fb7b46b0/llvm/lib/CodeGen/GlobalISel/Utils.cpp#L395

File: llvm/test/MachineVerifier/test_g_vscale.mir [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+name:            g_vscale
+body: |
+  bb.0:
+
+  %1:_(s32) = G_CONSTANT 4
+
+  ; CHECK: G_VSCALE operand must be cimm
+  %2:_(s32) = G_VSCALE %1
+
+  ; CHECK: G_VSCALE immediate cannot be zero
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: G_CONSTANT slightly differs:
```
def G_CONSTANT : GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins unknown:$imm);
  let hasSideEffects = false;
}
```
Note the `$imm`.

[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: @tschuett I changed `src` to `imm` but it doesn't seem to fix the problem

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: Failing in what way?

[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: The second case is reporting `G_VSCALE operand must be cimm`

[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: The second case is reporting `G_VSCALE operand must be cimm`

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/MachineVerifier/test_g_vscale.mir:13]: Oh right, I should have looked at more context. Yes, a literal 0 in the MIR is an isImm, not isCImm. You need "i32 0"

File: llvm/test/MachineVerifier/test_g_vscale.mir [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# RUN: not --crash llc -verify-machineinstrs -run-pass none -o /dev/null %s 2>&1 | FileCheck %s
+
+---
+name:            g_vscale
+body: |
+  bb.0:
+
+  %1:_(s32) = G_CONSTANT 4
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_g_vscale.mir:8]: This one should also be erroring

Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_vscale.mir]: Ahh, I see. Thanks for the help. I updated the test case.

**Review by arsenm — APPROVED** [review]

**Pull request merged**